### DISCUSSION TITLE
Best way to convert thrust to PWM for Crazyflie 2.1

### DISCUSSION THREAD

**Discussion by danialwaleed** [bitcraze/discussions#1222] [Q&A]
Hi there,

I am not 100% sure I have searched this forum thoroughly but I have not been able to get a good idea of how to convert thrust to PWM.

I am helping in developing a new controller for Crazyflie and what we want is to convert the output of the controller which is in thrust and torques directly to the motor commands.

I am finding it difficult to understand how that is done in Crazyflie. From what I understand the Crazyflie sets the PWM signal to the motors so how should I go about it. Any help will be appreciated.

a) I can convert the thrsut from N to grams if that helps

**Comment by knmcguire** [comment]
Hi!

So there are a couple of [controlModes](https://github.com/bitcraze/crazyflie-firmware/blob/f61da11d54b6d54c7fa746688e8e9be4edd73a29/src/modules/src/power_distribution_quadrotor.c#L122-L129) next to the controllers by itself 
* controlModeLegacy : These use altitude and thrust command directly for the motor mixing. This is a legacy mode and the PID/mellinger/indi controller outputs roll/pitch/yaw/thrust still. For PID this is put into the 'control_t' struct : https://github.com/bitcraze/crazyflie-firmware/blob/f61da11d54b6d54c7fa746688e8e9be4edd73a29/src/modules/src/controller/controller_pid.c#L133-L136
* controlModeForceTorque: These handle torques and thrust at powerdistribution. This is the output of the Berscianini controller
* controlModeForce: I persume this is supposed to use only force from the props but this has not been implemented yet.

So if you are making a new controller and it outputs torques and thrust than you should make sure that you define the control model to be controlModeForceTorque
